### PR TITLE
pkgsMusl.haskell.compiler.ghc8107Binary: Remove now-incorrect gmp dependency

### PR DESCRIPTION
**Pull request by nh2**
###### Motivation for this change

GHC HQ switched the musl bindists from gmp to `integer-simple`
with GHC >= 8.10.6, but this was not reflected in the nixpkgs update:

* commit 6f1242469a: ghc: 8.10.5-binary -> 8.10.7-binary
  From PR #135453

See also #130441.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by domenkozar** [comment]
I had no idea they switched.

**Comment by cdepillabout** [comment]
@nh2 How do you recommend this be tested?

Would building the following things fully test this?

**binary ghcs**

- `haskell.packages.ghc8102Binary.lens`: This would test that the gmp-linked non-musl ghc8102Binary can actually be used to build packages.
- `pkgsMusl.haskell.packages.ghc8102Binary.lens`: This would test that the gmp-linked musl `ghc8102Binary` can be used to build packages.
- `haskell.packages.ghc8107Binary.lens`: This would test that the gmp-linked non-musl ghc8107Binary can actually be used to build packages.
- `pkgsMusl.haskell.packages.ghc8107Binary.lens`: This would test that the non-gmp-linked musl `ghc8107Binary` can be used to build packages.

**ghc8107**

- `haskell.packages.ghc8107.lens`: This would test that a gmp-linked non-musl ghc8107 (which is built with the gmp-linked non-musl ghc865Binary) can be used to build packages.  *edit: I guess this shouldn't actually be affected by this PR at all*
- `pkgsMusl.haskell.packages.ghc8107.lens`: This would test that a gmp-linked musl ghc8107 (which is built with gmp-linked musl ghc8102BinaryMinimal) can be used to build packages.
- `haskell.packages.integer-simple.ghc8107.lens`: This would test that a non-gmp-linked non-musl ghc8107 (which is built with the gmp-linked non-musl ghc865Binary) can be used to build packages.  *edit: I guess this shouldn't actually be affected by this PR at all*
- `pkgsMusl.haskell.packages.integer-simple.ghc8107.lens`: This would test that a non-gmp-linked musl ghc8107 (which is built with gmp-linked musl ghc8102BinaryMinimal) can be used to build packages.

**ghc901**

- `haskell.packages.ghc901.lens`: This would test that a gmp-linked non-musl ghc901 (which is built with the gmp-linked non-musl ghc8102Binary) can be used to build packages.
- `pkgsMusl.haskell.packages.ghc901.lens`: This would test that a gmp-linked musl ghc901 (which is built with gmp-linked musl ghc8102Binary) can be used to build packages.
- `haskell.packages.integer-simple.ghc901.lens`: This would test that a non-gmp-linked non-musl ghc901 (which is built with the gmp-linked non-musl ghc865Binary) can be used to build packages.  *edit: I guess this shouldn't actually be affected by this PR at all*
- `pkgsMusl.haskell.packages.integer-simple.ghc901.lens`: This would test that a non-gmp-linked musl ghc901 (which is built with gmp-linked musl ghc8102Binary) can be used to build packages.

Is this about right?  Do the success of these derivations give a good indication as to whether this PR is correct?  Or is this overkill?

-----------------------------------------------------

The one thing the above derivation list is missing is something like the above `ghc901` section, but where `haskell.compiler.ghc901`, `pkgsMusl.haskell.compiler.ghc901`, `haskell.compiler.integer-simple.ghc901`, `pkgsMusl.haskell.compiler.integer-simple.ghc901` are built using `ghc8107Binary`.

Also, in `pkgs/top-level/haskell-packages.nix`, there is a note that starting from GHC 9, `integer-simple` and `integer-gmp` have been replaced by `ghc-bignum` with `native` and `gmp` backends.  Should these be tested as well??

**Comment by nh2** [comment]
@cdepillabout Personally I'd just build _all_ Haskell packages Hydra builds, and check if any of them regressed, if that's easy (I fail to remember how to do that though). However, all the packages you mentioned build fine for me (see below).

Recording here some failures and successes:

* **Failure:** `pkgsMusl.haskell.compiler.integer-simple.ghc884`
  ```
  /build/ghc27986_0/ghc_4.c:7:15: error:
       error: variable ‘stg_exports_FastString’ has initializer but incomplete type
          7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
            |               ^~~~~~~~~~~~~~~~~~
    |
  7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
    |               ^

  /build/ghc27986_0/ghc_4.c:7:61: error:
       error: ‘struct ForeignExportsList’ has no member named ‘exports’
          7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
            |                                                             ^~~~~~~
    |
  7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
    |                                                             ^

  /build/ghc27986_0/ghc_4.c:7:72: error:
       error: extra brace group at end of initializer
          7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
            |                                                                        ^
    |
  7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
    |                                                                        ^
  ```
* Success: `pkgsMusl.haskellPackages.hello`
* Success: `pkgsMusl.haskell.compiler.integer-simple.ghc8107`
* Success: `haskell.packages.ghc8102Binary.lens`
* Success: `pkgsMusl.haskell.packages.ghc8102Binary.lens`
* Success: `haskell.packages.ghc8107Binary.lens`
* Success: `pkgsMusl.haskell.packages.ghc8107Binary.lens`
* Success: `haskell.packages.ghc8107.lens`
* Success: `pkgsMusl.haskell.packages.ghc8107.lens`
* Success: `haskell.packages.integer-simple.ghc8107.lens`
* Success: `pkgsMusl.haskell.packages.integer-simple.ghc8107.lens`
* Success: `haskell.packages.ghc901.lens`
* Success: `pkgsMusl.haskell.packages.ghc901.lens`
* Success: `haskell.packages.integer-simple.ghc901.lens`
* Success: `pkgsMusl.haskell.packages.integer-simple.ghc901.lens`

**Review by cdepillabout — APPROVED** [review]

File: pkgs/development/compilers/ghc/8.10.7-binary.nix [hunk 1]
```diff
@@ -152,21 +157,21 @@ stdenv.mkDerivation rec {
...
+  #       the GHC 9 branch also switched from `gmp` to `integer-simple` via the
+  #       currently-open issue:
+  #           https://gitlab.haskell.org/ghc/ghc/-/issues/20059
+  #       and update this comment accordingly.
+
   nativeBuildInputs = [ perl ];
   propagatedBuildInputs =
     lib.optionals useLLVM [ llvmPackages.llvm ]
-    # Because musl bindists currently provide no way to tell where
-    # libgmp is (see not [musl bindists have no .buildinfo]), we need
```
Comment (sternenseemann) [inline pkgs/development/compilers/ghc/8.10.7-binary.nix:159]: Can the comment on musl bindists have no .buildinfo further down removed here as well (or clarified wrt musl)?

**Comment by sternenseemann** [comment]
Maybe let's close this in favor of #138524 for simplicities' sake? Would like to get all changes in this rotation.

File: pkgs/development/compilers/ghc/8.10.2-binary.nix [hunk 2]
```diff
@@ -140,20 +143,35 @@ stdenv.mkDerivation rec {
 
   src = fetchurl binDistUsed.src;
 
+  # Note that for GHC 8.10 versions <= 8.10.5, the GHC HQ musl bindist
```
[the remark this answers is not in the document]
Comment (nh2) [reply pkgs/development/compilers/ghc/8.10.2-binary.nix:146]: Done.

File: pkgs/development/compilers/ghc/8.10.7-binary.nix (hunk 1 above)
Comment (nh2) [reply pkgs/development/compilers/ghc/8.10.7-binary.nix:159]: Thanks, I have fixed the dangling reference.

**Comment by nh2** [comment]
> Maybe let's close this in favor of #138524 for simplicity's sake? Would like to get all changes in this rotation.

Sure, can do either.

I made this one separately because it is much easier, as it touches only musl; the other one touches all compilers.

**Comment by sternenseemann** [comment]
I'd just do the set rebuild all in one go, actually.

**Comment by nh2** [comment]
> * **Failure:** `pkgsMusl.haskell.compiler.integer-simple.ghc884`
>   ```
>   /build/ghc27986_0/ghc_4.c:7:15: error:
>        error: variable ‘stg_exports_FastString’ has initializer but incomplete type
>           7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
>             |               ^~~~~~~~~~~~~~~~~~
>     |
>   7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
>     |               ^
>   
>   /build/ghc27986_0/ghc_4.c:7:61: error:
>        error: ‘struct ForeignExportsList’ has no member named ‘exports’
>           7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
>             |                                                             ^~~~~~~
>     |
>   7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
>     |                                                             ^
>   
>   /build/ghc27986_0/ghc_4.c:7:72: error:
>        error: extra brace group at end of initializer
>           7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
>             |                                                                        ^
>     |
>   7 | static struct ForeignExportsList stg_exports_FastString = {.exports =  {}, .n_entries =  0};
>     |                                                                        ^
>   ```

@sternenseemann Investigating this deeper, I found some odd results:

`let mc = pkgsMusl.haskell.compiler` in:

1. `ghc8102BinaryMinimal` **succeeds** at `mc.ghc884` fine.
2. `ghc8107BinaryMinimal` **fails** at `mc.ghc884`, with the above C error,
   so the `.2` -> `.7` bindist regressed that
3. `ghc8102BinaryMinimal` **fails** at `mc.integer-simple.ghc884`, with the above C error
4. `ghc8102BinaryMinimal` **fails** at `mc.integer-simple.ghc884` with the old expected error
   `ld: cannot find -lgmp`

**Pull request merged**